File: day4-orchestration-evals-cicd/lab5-1-handoff/store.py
"""
The shared state store: the ONLY thing the stages have in common.

Stages never call each other. Each reads what it needs from here and writes its
result back. That is what makes the pipeline checkpointed (a crash loses at
most the stage in flight), resumable (re-running skips finished stages),
inspectable (every hand-off is a row you can read), and gateable (a human
decision is a row too, with a name and a reason).

SQLite, standard library. The same shape works on Postgres or DynamoDB.
"""
import hashlib, json, sqlite3, threading, uuid
from datetime import datetime, timedelta, timezone

IST = timezone(timedelta(hours=5, minutes=30))
def now():
    return datetime.now(IST).isoformat(timespec="seconds")

SCHEMA = """
create table if not exists runs(id text primary key, account_id text, question text,
    status text, created_at text, updated_at text);
create table if not exists stages(run_id text, name text, status text, attempt integer,
    output text, error text, cost_usd real, tool_calls integer, started_at text, finished_at text,
    primary key(run_id, name));
create table if not exists approvals(run_id text primary key, decision text, approver text,
    reason text, override integer, at text, proposal_sha text);
create table if not exists operations(run_id text primary key, op_id text, action text,
    payload text, status text, response text, created_at text, updated_at text);
"""

# Run statuses, in order. The pipeline moves forward only.
#   created -> investigated -> reviewed -> awaiting_approval -> approved -> applied
#                                      \-> needs_rework        \-> rejected
#   apply can end in: applied | apply_failed | outcome_unknown (resume retries safely)
TERMINAL = ("applied", "rejected", "no_change")

class Conflict(RuntimeError):
    """Two writers raced for a row only one of them may create (e.g. two people deciding one run)."""

class Store:
    def __init__(self, path):
        self.db = sqlite3.connect(path, check_same_thread=False)
        self.db.row_factory = sqlite3.Row
        self.db.executescript(SCHEMA)
        cols = {r[1] for r in self.db.execute("pragma table_info(approvals)")}
        if "proposal_sha" not in cols:            # a runs.sqlite from before this column existed
            self.db.execute("alter table approvals add column proposal_sha text")
        self.lock = threading.Lock()

    # ---- runs
    def create_run(self, account_id, question):
        rid = uuid.uuid4().hex[:10]
        with self.lock:
            self.db.execute("insert into runs values(?,?,?,?,?,?)", (rid, account_id, question, "created", now(), now()))
            self.db.commit()
        return rid

    def run(self, rid):
        r = self.db.execute("select * from runs where id=?", (rid,)).fetchone()
        if not r:
            raise KeyError(f"no run {rid}")
        return dict(r)

    def runs(self, limit=20):
        return [dict(r) for r in self.db.execute("select * from runs order by created_at desc limit ?", (limit,))]

    def set_status(self, rid, status):
        with self.lock:
            self.db.execute("update runs set status=?, updated_at=? where id=?", (status, now(), rid))
            self.db.commit()

    # ---- stages: the checkpoint
    def stage(self, rid, name):
        r = self.db.execute("select * from stages where run_id=? and name=?", (rid, name)).fetchone()
        if not r:
            return None
        d = dict(r)
        d["output"] = json.loads(d["output"]) if d["output"] else None
        return d

    def stage_started(self, rid, name):
        prev = self.stage(rid, name)
        attempt = (prev["attempt"] + 1) if prev else 1
        spent = prev["cost_usd"] if prev else 0.0          # failed attempts were paid for too
        with self.lock:
            self.db.execute("insert or replace into stages values(?,?,?,?,?,?,?,?,?,?)",
                            (rid, name, "running", attempt, None, None, spent, 0, now(), None))
            self.db.commit()
        return attempt

    def stage_done(self, rid, name, output, cost_usd, tool_calls):
        with self.lock:
            self.db.execute("update stages set status='done', output=?, cost_usd=cost_usd+?, tool_calls=?, finished_at=?, error=null "
                            "where run_id=? and name=?", (json.dumps(output), cost_usd, tool_calls, now(), rid, name))
            self.db.commit()

    def stage_failed(self, rid, name, error, cost_usd=0.0):
        with self.lock:
            self.db.execute("update stages set status='failed', error=?, cost_usd=cost_usd+?, finished_at=? where run_id=? and name=?",
                            (str(error)[:500], cost_usd, now(), rid, name))
            self.db.commit()

    # ---- the human gate
    def approval(self, rid):
        r = self.db.execute("select * from approvals where run_id=?", (rid,)).fetchone()
        return dict(r) if r else None

    def proposal_sha(self, rid):
        """Fingerprint of the exact change a human is deciding on."""
        st = self.stage(rid, "investigate")
        change = (st or {}).get("output", {}) and st["output"].get("proposed_change")
        return hashlib.sha256(json.dumps(change, sort_keys=True).encode()).hexdigest()

    def record_decision(self, rid, decision, approver, reason, override=False):
        # The decision is bound to the proposal as it was when the human saw it.
        with self.lock:
            try:
                self.db.execute("insert into approvals(run_id, decision, approver, reason, override, at, proposal_sha) "
                                "values(?,?,?,?,?,?,?)",
                                (rid, decision, approver, reason, int(override), now(), self.proposal_sha(rid)))
            except sqlite3.IntegrityError:        # the primary key is the real "one decision" guarantee
                raise Conflict(f"run {rid} was already decided") from None
            self.db.commit()

    # ---- the write: its operation id is stored BEFORE it is sent
    def operation(self, rid):
        r = self.db.execute("select * from operations where run_id=?", (rid,)).fetchone()
        if not r:
            return None
        d = dict(r)
        d["payload"] = json.loads(d["payload"])
        return d

    def record_operation(self, rid, op_id, action, payload):
        with self.lock:
            self.db.execute("insert into operations values(?,?,?,?,?,?,?,?)",
                            (rid, op_id, action, json.dumps(payload), "pending", None, now(), now()))
            self.db.commit()

    def operation_result(self, rid, status, response):
        with self.lock:
            self.db.execute("update operations set status=?, response=?, updated_at=? where run_id=?",
                            (status, json.dumps(response)[:4000], now(), rid))
            self.db.commit()

    def cost(self, rid):
        return round(self.db.execute("select coalesce(sum(cost_usd),0) from stages where run_id=?", (rid,)).fetchone()[0], 4)

File: day4-orchestration-evals-cicd/java/lab51/src/main/java/com/airamatrix/day4/lab51/Store.java
package com.airamatrix.day4.lab51;

import java.nio.charset.StandardCharsets;
import java.security.MessageDigest;
import java.security.NoSuchAlgorithmException;
import java.sql.Connection;
import java.sql.DriverManager;
import java.sql.PreparedStatement;
import java.sql.ResultSet;
import java.sql.SQLException;
import java.sql.Statement;
import java.time.OffsetDateTime;
import java.time.ZoneOffset;
import java.time.format.DateTimeFormatter;
import java.util.ArrayList;
import java.util.HexFormat;
import java.util.HashSet;
import java.util.List;
import java.util.Set;
import java.util.UUID;

import com.airamatrix.day4.common.Contracts;
import com.fasterxml.jackson.core.JsonProcessingException;
import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.node.NullNode;

/**
 * The shared state store: the ONLY thing the stages have in common. Port of
 * lab5-1-handoff/store.py - same tables, same columns, same semantics, so a runs.sqlite
 * written by one can be read by the other.
 *
 * Stages never call each other. Each reads what it needs from here and writes its result back.
 * That makes the pipeline checkpointed (a crash loses at most the stage in flight), resumable
 * (re-running skips finished stages), inspectable (every hand-off is a row you can read) and
 * gateable (a human decision is a row too, with a name and a reason).
 *
 * SQLite over plain JDBC, one connection, every method synchronized (the Python lock).
 */
public class Store implements AutoCloseable {

    static final ZoneOffset IST = ZoneOffset.ofHoursMinutes(5, 30);
    private static final DateTimeFormatter ISO_SECONDS = DateTimeFormatter.ofPattern("yyyy-MM-dd'T'HH:mm:ssxxx");

    static String now() {
        return OffsetDateTime.now(IST).format(ISO_SECONDS);
    }

    static final String[] SCHEMA = {
        "create table if not exists runs(id text primary key, account_id text, question text, "
            + "status text, created_at text, updated_at text)",
        "create table if not exists stages(run_id text, name text, status text, attempt integer, "
            + "output text, error text, cost_usd real, tool_calls integer, started_at text, finished_at text, "
            + "primary key(run_id, name))",
        "create table if not exists approvals(run_id text primary key, decision text, approver text, "
            + "reason text, override integer, at text, proposal_sha text)",
        "create table if not exists operations(run_id text primary key, op_id text, action text, "
            + "payload text, status text, response text, created_at text, updated_at text)",
    };

    // Run statuses, in order. The pipeline moves forward only.
    //   created -> investigated -> reviewed -> awaiting_approval -> approved -> applied
    //                                      \-> needs_rework        \-> rejected
    //   apply can end in: applied | apply_failed | outcome_unknown (apply again retries safely)
    public static final List<String> TERMINAL = List.of("applied", "rejected", "no_change");

    /** Python raises KeyError("no run X"); the CLI prints it the same way: refused: 'no run X'. */
    public static class NoSuchRun extends RuntimeException {
        public NoSuchRun(String rid) { super("no run " + rid); }
    }

    /** Two writers raced for a row only one of them may create (e.g. two people deciding one run). */
    public static class Conflict extends RuntimeException {
        public Conflict(String message) { super(message); }
    }

    public record RunRow(String id, String accountId, String question, String status, String createdAt, String updatedAt) {}

    public record StageRow(String runId, String name, String status, int attempt, JsonNode output, String error,
                           double costUsd, int toolCalls, String startedAt, String finishedAt) {}

    public record ApprovalRow(String runId, String decision, String approver, String reason, int override,
                              String at, String proposalSha) {}

    public record OperationRow(String runId, String opId, String action, JsonNode payload, String status,
                               String response, String createdAt, String updatedAt) {}

    private final Connection db;
    public final String path;

    public Store(String path) {
        this.path = path;
        try {
            db = DriverManager.getConnection("jdbc:sqlite:" + path);
            try (Statement st = db.createStatement()) {
                st.execute("pragma busy_timeout = 5000");       // the CLI and a server may share the file
                for (String ddl : SCHEMA) st.execute(ddl);
                Set<String> cols = new HashSet<>();
                try (ResultSet rs = st.executeQuery("pragma table_info(approvals)")) {
                    while (rs.next()) cols.add(rs.getString(2));
                }
                if (!cols.contains("proposal_sha")) {           // a runs.sqlite from before this column existed
                    st.execute("alter table approvals add column proposal_sha text");
                }
            }
        } catch (SQLException e) {
            throw new IllegalStateException("cannot open " + path + ": " + e.getMessage(), e);
        }
    }

    @Override
    public synchronized void close() {
        try { db.close(); } catch (SQLException ignored) { /* closing */ }
    }

    // ---- runs
    public synchronized String createRun(String accountId, String question) {
        String rid = UUID.randomUUID().toString().replace("-", "").substring(0, 10);
        update("insert into runs values(?,?,?,?,?,?)", rid, accountId, question, "created", now(), now());
        return rid;
    }

    public synchronized RunRow run(String rid) {
        List<RunRow> rows = query("select * from runs where id=?", this::runRow, rid);
        if (rows.isEmpty()) throw new NoSuchRun(rid);
        return rows.get(0);
    }

    public synchronized List<RunRow> runs(int limit) {
        return query("select * from runs order by created_at desc limit ?", this::runRow, limit);
    }

    public List<RunRow> runs() { return runs(20); }

    public synchronized void setStatus(String rid, String status) {
        update("update runs set status=?, updated_at=? where id=?", status, now(), rid);
    }

    // ---- stages: the checkpoint
    public synchronized StageRow stage(String rid, String name) {
        List<StageRow> rows = query("select * from stages where run_id=? and name=?", this::stageRow, rid, name);
        return rows.isEmpty() ? null : rows.get(0);
    }

    public synchronized int stageStarted(String rid, String name) {
        StageRow prev = stage(rid, name);
        int attempt = prev != null ? prev.attempt() + 1 : 1;
        double spent = prev != null ? prev.costUsd() : 0.0;          // failed attempts were paid for too
        update("insert or replace into stages values(?,?,?,?,?,?,?,?,?,?)",
                rid, name, "running", attempt, null, null, spent, 0, now(), null);
        return attempt;
    }

    public synchronized void stageDone(String rid, String name, JsonNode output, double costUsd, int toolCalls) {
        update("update stages set status='done', output=?, cost_usd=cost_usd+?, tool_calls=?, finished_at=?, error=null "
                + "where run_id=? and name=?", PyJson.dumps(output), costUsd, toolCalls, now(), rid, name);
    }

    public synchronized void stageFailed(String rid, String name, String error, double costUsd) {
        String e = String.valueOf(error);
        update("update stages set status='failed', error=?, cost_usd=cost_usd+?, finished_at=? where run_id=? and name=?",
                e.length() > 500 ? e.substring(0, 500) : e, costUsd, now(), rid, name);
    }

    // ---- the human gate
    public synchronized ApprovalRow approval(String rid) {
        List<ApprovalRow> rows = query("select * from approvals where run_id=?", rs -> new ApprovalRow(
                rs.getString("run_id"), rs.getString("decision"), rs.getString("approver"), rs.getString("reason"),
                rs.getInt("override"), rs.getString("at"), rs.getString("proposal_sha")), rid);
        return rows.isEmpty() ? null : rows.get(0);
    }

    /** Fingerprint of the exact change a human is deciding on: sha256(json.dumps(change, sort_keys=True)). */
    public synchronized String proposalSha(String rid) {
        StageRow st = stage(rid, "investigate");
        JsonNode change;
        if (st == null) change = Contracts.object();                           // Python: {}
        else if (st.output() == null || st.output().isEmpty()) change = st.output() == null ? NullNode.instance : st.output();
        else change = st.output().has("proposed_change") ? st.output().get("proposed_change") : NullNode.instance;
        try {
            byte[] h = MessageDigest.getInstance("SHA-256").digest(PyJson.dumpsSorted(change).getBytes(StandardCharsets.UTF_8));
            return HexFormat.of().formatHex(h);
        } catch (NoSuchAlgorithmException e) {
            throw new IllegalStateException(e);
        }
    }

    public synchronized void recordDecision(String rid, String decision, String approver, String reason, boolean override) {
        // The decision is bound to the proposal as it was when the human saw it.
        try (PreparedStatement ps = db.prepareStatement("insert into approvals(run_id, decision, approver, reason, override, at, proposal_sha) "
                + "values(?,?,?,?,?,?,?)")) {
            bind(ps, rid, decision, approver, reason, override ? 1 : 0, now(), proposalSha(rid));
            ps.executeUpdate();
        } catch (SQLException e) {
            // the primary key is the real "one decision" guarantee
            if (String.valueOf(e.getMessage()).contains("SQLITE_CONSTRAINT")) {
                throw new Conflict("run " + rid + " was already decided");
            }
            throw new IllegalStateException(e);
        }
    }

    // ---- the write: its operation id is stored BEFORE it is sent
    public synchronized OperationRow operation(String rid) {
        List<OperationRow> rows = query("select * from operations where run_id=?", rs -> new OperationRow(
                rs.getString("run_id"), rs.getString("op_id"), rs.getString("action"), parse(rs.getString("payload")),
                rs.getString("status"), rs.getString("response"), rs.getString("created_at"), rs.getString("updated_at")), rid);
        return rows.isEmpty() ? null : rows.get(0);
    }

    public synchronized void recordOperation(String rid, String opId, String action, JsonNode payload) {
        update("insert into operations values(?,?,?,?,?,?,?,?)",
                rid, opId, action, PyJson.dumps(payload), "pending", null, now(), now());
    }

    public synchronized void operationResult(String rid, String status, JsonNode response) {
        String r = PyJson.dumps(response);
        update("update operations set status=?, response=?, updated_at=? where run_id=?",
                status, r.length() > 4000 ? r.substring(0, 4000) : r, now(), rid);
    }

    public synchronized double cost(String rid) {
        List<Double> v = query("select coalesce(sum(cost_usd),0) from stages where run_id=?", rs -> rs.getDouble(1), rid);
        return PyJson.round4(v.get(0));
    }

    // ---- plumbing
    private RunRow runRow(ResultSet rs) throws SQLException {
        return new RunRow(rs.getString("id"), rs.getString("account_id"), rs.getString("question"),
                rs.getString("status"), rs.getString("created_at"), rs.getString("updated_at"));
    }

    private StageRow stageRow(ResultSet rs) throws SQLException {
        return new StageRow(rs.getString("run_id"), rs.getString("name"), rs.getString("status"), rs.getInt("attempt"),
                parse(rs.getString("output")), rs.getString("error"), rs.getDouble("cost_usd"), rs.getInt("tool_calls"),
                rs.getString("started_at"), rs.getString("finished_at"));
    }

    static JsonNode parse(String s) {
        if (s == null || s.isEmpty()) return null;
        try {
            return Contracts.JSON.readTree(s);
        } catch (JsonProcessingException e) {
            throw new IllegalStateException("bad JSON in store: " + e.getOriginalMessage());
        }
    }

    interface Mapper<T> { T map(ResultSet rs) throws SQLException; }

    private <T> List<T> query(String sql, Mapper<T> mapper, Object... args) {
        try (PreparedStatement ps = db.prepareStatement(sql)) {
            bind(ps, args);
            try (ResultSet rs = ps.executeQuery()) {
                List<T> out = new ArrayList<>();
                while (rs.next()) out.add(mapper.map(rs));
                return out;
            }
        } catch (SQLException e) {
            throw new IllegalStateException(e.getMessage(), e);
        }
    }

    private void update(String sql, Object... args) {
        try (PreparedStatement ps = db.prepareStatement(sql)) {
            bind(ps, args);
            ps.executeUpdate();
        } catch (SQLException e) {
            throw new IllegalStateException(e.getMessage(), e);
        }
    }

    private static void bind(PreparedStatement ps, Object... args) throws SQLException {
        for (int i = 0; i < args.length; i++) ps.setObject(i + 1, args[i]);
    }
}
